### PR TITLE
Refactored MGLUserLocationAnnotationView (redux)

### PR DESCRIPTION
**Pull request by friedbunny**
This is a replacement PR for https://github.com/mapbox/mapbox-gl-native/pull/5573 that targets `master`.

Additionally, I’ve rewritten https://github.com/mapbox/mapbox-gl-native/pull/5816 so that the hit testing layer can be set by any subclass of `MGLUserLocationAnnotationView`.

/cc @frederoni @1ec5 @incanus @boundsj

### CODE REVIEW

File: platform/ios/src/MGLMapView.mm [hunk 1]
```diff
@@ -1397,15 +1397,22 @@ - (void)handleSingleTapGesture:(UITapGestureRecognizer *)singleTap
...
 
-    CALayer *hitLayer = self.userLocationVisible ? [self.userLocationAnnotationView.layer.presentationLayer hitTest:tapPoint] : nil;
-    if (hitLayer && hitLayer != self.userLocationAnnotationView.haloLayer.presentationLayer)
+    if (self.userLocationVisible)
     {
-        if ( ! _userLocationAnnotationIsSelected)
+
+        CGPoint tapPointForUserLocation = self.userLocationAnnotationView.hitTestLayer.superlayer ? [singleTap locationInView:self.userLocationAnnotationView] : tapPoint;
+
+        CALayer *hitLayer = self.userLocationVisible ? [self.userLocationAnnotationView.hitTestLayer hitTest:tapPointForUserLocation] : nil;
```
Comment (friedbunny) [inline platform/ios/src/MGLMapView.mm]: This doesn’t need to check again for `self.userLocationVisible`, will remove.

**Comment by friedbunny** [comment]
Saw some weird behavior just now that I don’t have time to investigate, but the user dot wasn’t being updated/deselected when another annotation became selected. E.g., tapping another annotation while in course tracking mode would leave the puck on screen until the tracking mode was manually changed again.

Will investigate next week.

**Comment by friedbunny** [comment]
Rebased again after v3.3.4 was merged back to master, which included a backport of 7aec17d in https://github.com/mapbox/mapbox-gl-native/pull/5894.

**Comment by friedbunny** [comment]
380d54a9d5106fb957455c4376e086634ff05f39 fixes the issue where the annotation was not updating immediately. The `didUpdateUserLocation:` method was changed to `update`, which is again called from `updateUserLocationAnnotationViewAnimatedWithDuration:`.

The user location annotation needs updates more frequently once per location event — it needs to respond to viewport changes and user tracking mode changes.

Very open to naming suggestions for `update`.

**Comment by friedbunny** [comment]
Updated the docs a bit. Is the `@warning` too much?

screen shot 2016-08-12 at 10 41 26 pm

File: platform/ios/app/MBXUserLocationAnnotationView.m [hunk 2]
```diff
@@ -0,0 +1,172 @@
+#import "MBXUserLocationAnnotationView.h"
```
Comment (boundsj) [inline platform/ios/app/MBXUserLocationAnnotationView.m]: Not: import not required

File: platform/ios/app/MBXUserLocationAnnotationView.m [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+#import <Mapbox/Mapbox.h>
```
Comment (boundsj) [inline platform/ios/app/MBXUserLocationAnnotationView.m]: Nit import not required

File: platform/ios/app/MBXViewController.m [hunk 4]
```diff
@@ -699,6 +700,13 @@ - (void)dealloc
 
 - (MGLAnnotationView *)mapView:(MGLMapView *)mapView viewForAnnotation:(id<MGLAnnotation>)annotation
 {
+    if (annotation == mapView.userLocation)
+    {
+        //MBXUserLocationAnnotationView *annotationView = [[MBXUserLocationAnnotationView alloc] initWithFrame:CGRectZero];
```
Comment (boundsj) [inline platform/ios/app/MBXViewController.m]: Maybe we should add an option in the app UI to exercise the custom view?

Comment (friedbunny) [inline platform/ios/app/MBXViewController.m]: We should!

**Comment by incanus** [comment]
> The didUpdateUserLocation: method has been changed to update, which is again called from `-[MGLMapView updateUserLocationAnnotationViewAnimatedWithDuration:]`.

Can you say more about this @friedbunny? `didUpdateUserLocation:` is akin to how MapKit does things.

**Comment by incanus** [comment]
Aha! Thank you.

**Comment by friedbunny** [comment]
Rebased again — I’m going to aim for merging this afternoon.

**Pull request merged**